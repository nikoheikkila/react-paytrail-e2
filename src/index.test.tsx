import React from 'react'
import { render } from '@testing-library/react'
import { Form } from '.'

const getURLs = (baseURI: string): URLSet => ({
  success: new URL(`${baseURI}/success`),
  cancel: new URL(`${baseURI}/cancel`),
  notify: new URL(`${baseURI}/notify`),
})

test('Form renders with minimum required fields', async () => {
  const merchant: Merchant = {
    id: '1',
    secret: '123',
  }

  const orderNumber = '1'
  const urls = getURLs('https://example.com')

  const customer: Customer = {
    firstName: 'Alice',
    lastName: 'Angel',
  }

  const product: Product = {
    id: '123',
    title: 'Test Product',
    price: 100,
  }

  const { container } = render(<Form merchant={merchant} orderNumber={orderNumber} customer={customer} products={[product]} urls={urls} />)

  expect(container).toMatchSnapshot()
  expect(container.querySelectorAll('input')).toHaveLength(21)
  expect(container.querySelector('input[name=MERCHANT_ID]')).toHaveProperty('value', merchant.id)
  expect(container.querySelector('input[name=ORDER_NUMBER]')).toHaveProperty('value', orderNumber)
  expect(container.querySelector('input[name=PAYER_PERSON_FIRSTNAME]')).toHaveProperty('value', customer.firstName)
  expect(container.querySelector('input[name=PAYER_PERSON_LASTNAME]')).toHaveProperty('value', customer.lastName)
  expect(container.querySelector('input[name^=ITEM_ID')).toHaveProperty('value', product.id)
  expect(container.querySelector('input[name^=ITEM_TITLE')).toHaveProperty('value', product.title)
  expect(container.querySelector('input[name^=ITEM_UNIT_PRICE')).toHaveProperty('value', '100.00')
})

test('Form renders with all the fields', async () => {
  const merchant: Merchant = {
    id: '1',
    secret: '123',
  }

  const className = 'e2-test-form'
  const orderNumber = '1'
  const paymentMethods = [1, 2]
  const reference = 'RF123456789'
  const currency = 'EUR'
  const locale = 'fi_FI'
  const algorithm = 'sha256'
  const includeVat = false

  const expiresAt = '2021-01-01'
  const urls = getURLs('https://example.com')

  const customer: Customer = {
    firstName: 'Alice',
    lastName: 'Angel',
    phone: '040123456',
    email: 'alice@example.org',
    company: 'Acme Ltd',
    address: {
      street: 'Test Street 1',
      postalCode: '00100',
      town: 'Helsinki',
      country: 'FI',
    },
  }

  const product: Product = {
    id: '123',
    title: 'Test Product',
    price: 100,
    discount: '10',
    vat: '24',
    quantity: '1',
    type: '1',
  }

  const messages = {
    merchantPanel: "This is a message for the Merchant's Panel",
    payer: 'This is a message for the customer',
    paymentMethod: 'This is a message for the payment method',
  }

  const { container } = render(
    <Form
      className={className}
      debug
      merchant={merchant}
      orderNumber={orderNumber}
      urls={urls}
      customer={customer}
      products={[product]}
      currency={currency}
      locale={locale}
      reference={reference}
      paymentMethods={paymentMethods}
      includeVat={includeVat}
      algorithm={algorithm}
      messages={messages}
      expiresAt={expiresAt}
    />,
  )

  const expectedInputFields = 34
  const expectedDebugFields = expectedInputFields - 1 // AUTHCODE is not shown in debug

  expect(container).toMatchSnapshot()
  expect(container.querySelectorAll('input')).toHaveLength(expectedInputFields)
  expect(container.querySelectorAll('ul > li')).toHaveLength(expectedDebugFields)
  expect(container.querySelector('input[name=MERCHANT_ID')).toHaveProperty('value', merchant.id)
  expect(container.querySelector('input[name=ORDER_NUMBER')).toHaveProperty('value', orderNumber)
  expect(container.querySelector('input[name=PAYMENT_METHODS')).toHaveProperty('value', '1,2')
  expect(container.querySelector('input[name=REFERENCE_NUMBER')).toHaveProperty('value', reference)
  expect(container.querySelector('input[name=CURRENCY')).toHaveProperty('value', currency)
  expect(container.querySelector('input[name=LOCALE')).toHaveProperty('value', locale)
  expect(container.querySelector('input[name=VAT_IS_INCLUDED')).toHaveProperty('value', '0')
  expect(container.querySelector('input[name=EXPIRATION_FOR_PAYMENT_CREATION')).toHaveProperty('value', expiresAt)
  expect(container.querySelector('input[name=URL_SUCCESS')).toHaveProperty('value', urls.success.href)
  expect(container.querySelector('input[name=URL_CANCEL')).toHaveProperty('value', urls.cancel.href)
  expect(container.querySelector('input[name=URL_NOTIFY')).toHaveProperty('value', urls.notify?.href)
  expect(container.querySelector('input[name=PAYER_PERSON_FIRSTNAME]')).toHaveProperty('value', customer.firstName)
  expect(container.querySelector('input[name=PAYER_PERSON_LASTNAME]')).toHaveProperty('value', customer.lastName)
  expect(container.querySelector('input[name=PAYER_PERSON_PHONE]')).toHaveProperty('value', customer.phone)
  expect(container.querySelector('input[name=PAYER_PERSON_EMAIL]')).toHaveProperty('value', customer.email)
  expect(container.querySelector('input[name=PAYER_COMPANY_NAME]')).toHaveProperty('value', customer.company)
  expect(container.querySelector('input[name=PAYER_PERSON_ADDR_STREET]')).toHaveProperty('value', customer.address?.street)
  expect(container.querySelector('input[name=PAYER_PERSON_ADDR_POSTAL_CODE]')).toHaveProperty('value', customer.address?.postalCode)
  expect(container.querySelector('input[name=PAYER_PERSON_ADDR_TOWN]')).toHaveProperty('value', customer.address?.town)
  expect(container.querySelector('input[name=PAYER_PERSON_ADDR_COUNTRY]')).toHaveProperty('value', customer.address?.country)
  expect(container.querySelector('input[name^=ITEM_ID]')).toHaveProperty('value', product.id)
  expect(container.querySelector('input[name^=ITEM_TITLE]')).toHaveProperty('value', product.title)
  expect(container.querySelector('input[name^=ITEM_UNIT_PRICE]')).toHaveProperty('value', '100.00')
  expect(container.querySelector('input[name^=ITEM_VAT_PERCENT]')).toHaveProperty('value', product.vat)
  expect(container.querySelector('input[name^=ITEM_DISCOUNT_PERCENT]')).toHaveProperty('value', product.discount)
  expect(container.querySelector('input[name^=ITEM_TYPE]')).toHaveProperty('value', product.type)
  expect(container.querySelector('input[name=MSG_UI_MERCHANT_PANEL]')).toHaveProperty('value', messages.merchantPanel)
  expect(container.querySelector('input[name=MSG_UI_PAYMENT_METHOD]')).toHaveProperty('value', messages.paymentMethod)
  expect(container.querySelector('input[name=MSG_SETTLEMENT_PAYER]')).toHaveProperty('value', messages.payer)
})
