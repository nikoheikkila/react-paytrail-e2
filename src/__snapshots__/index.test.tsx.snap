// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders with all the fields 1`] = `
<div>
  <div>
    <ul>
      <li>
        <strong>
          MERCHANT_ID
          :
        </strong>
        <br />
        <code>
           1
        </code>
      </li>
      <li>
        <strong>
          ORDER_NUMBER
          :
        </strong>
        <br />
        <code>
           1
        </code>
      </li>
      <li>
        <strong>
          PARAMS_IN
          :
        </strong>
        <br />
        <code>
           MERCHANT_ID,ORDER_NUMBER,PARAMS_IN,LOCALE,CURRENCY,REFERENCE_NUMBER,EXPIRATION_FOR_PAYMENT_CREATION,PAYMENT_METHODS,VAT_IS_INCLUDED,ALG,URL_SUCCESS,URL_CANCEL,URL_NOTIFY,ITEM_TITLE[0],ITEM_ID[0],ITEM_QUANTITY[0],ITEM_UNIT_PRICE[0],ITEM_VAT_PERCENT[0],ITEM_DISCOUNT_PERCENT[0],ITEM_TYPE[0],PAYER_PERSON_PHONE,PAYER_PERSON_EMAIL,PAYER_PERSON_FIRSTNAME,PAYER_PERSON_LASTNAME,PAYER_COMPANY_NAME,PAYER_PERSON_ADDR_STREET,PAYER_PERSON_ADDR_POSTAL_CODE,PAYER_PERSON_ADDR_TOWN,PAYER_PERSON_ADDR_COUNTRY,MSG_UI_MERCHANT_PANEL,MSG_SETTLEMENT_PAYER,MSG_UI_PAYMENT_METHOD,PARAMS_OUT
        </code>
      </li>
      <li>
        <strong>
          LOCALE
          :
        </strong>
        <br />
        <code>
           fi_FI
        </code>
      </li>
      <li>
        <strong>
          CURRENCY
          :
        </strong>
        <br />
        <code>
           EUR
        </code>
      </li>
      <li>
        <strong>
          REFERENCE_NUMBER
          :
        </strong>
        <br />
        <code>
           RF123456789
        </code>
      </li>
      <li>
        <strong>
          EXPIRATION_FOR_PAYMENT_CREATION
          :
        </strong>
        <br />
        <code>
           2021-01-01
        </code>
      </li>
      <li>
        <strong>
          PAYMENT_METHODS
          :
        </strong>
        <br />
        <code>
           1,2
        </code>
      </li>
      <li>
        <strong>
          VAT_IS_INCLUDED
          :
        </strong>
        <br />
        <code>
           0
        </code>
      </li>
      <li>
        <strong>
          ALG
          :
        </strong>
        <br />
        <code>
           1
        </code>
      </li>
      <li>
        <strong>
          URL_SUCCESS
          :
        </strong>
        <br />
        <code>
           https://example.com/success
        </code>
      </li>
      <li>
        <strong>
          URL_CANCEL
          :
        </strong>
        <br />
        <code>
           https://example.com/cancel
        </code>
      </li>
      <li>
        <strong>
          URL_NOTIFY
          :
        </strong>
        <br />
        <code>
           https://example.com/notify
        </code>
      </li>
      <li>
        <strong>
          ITEM_TITLE[0]
          :
        </strong>
        <br />
        <code>
           Test Product
        </code>
      </li>
      <li>
        <strong>
          ITEM_ID[0]
          :
        </strong>
        <br />
        <code>
           123
        </code>
      </li>
      <li>
        <strong>
          ITEM_QUANTITY[0]
          :
        </strong>
        <br />
        <code>
           1
        </code>
      </li>
      <li>
        <strong>
          ITEM_UNIT_PRICE[0]
          :
        </strong>
        <br />
        <code>
           100.00
        </code>
      </li>
      <li>
        <strong>
          ITEM_VAT_PERCENT[0]
          :
        </strong>
        <br />
        <code>
           24
        </code>
      </li>
      <li>
        <strong>
          ITEM_DISCOUNT_PERCENT[0]
          :
        </strong>
        <br />
        <code>
           10
        </code>
      </li>
      <li>
        <strong>
          ITEM_TYPE[0]
          :
        </strong>
        <br />
        <code>
           1
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_PHONE
          :
        </strong>
        <br />
        <code>
           040123456
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_EMAIL
          :
        </strong>
        <br />
        <code>
           alice@example.org
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_FIRSTNAME
          :
        </strong>
        <br />
        <code>
           Alice
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_LASTNAME
          :
        </strong>
        <br />
        <code>
           Angel
        </code>
      </li>
      <li>
        <strong>
          PAYER_COMPANY_NAME
          :
        </strong>
        <br />
        <code>
           Acme Ltd
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_ADDR_STREET
          :
        </strong>
        <br />
        <code>
           Test Street 1
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_ADDR_POSTAL_CODE
          :
        </strong>
        <br />
        <code>
           00100
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_ADDR_TOWN
          :
        </strong>
        <br />
        <code>
           Helsinki
        </code>
      </li>
      <li>
        <strong>
          PAYER_PERSON_ADDR_COUNTRY
          :
        </strong>
        <br />
        <code>
           FI
        </code>
      </li>
      <li>
        <strong>
          MSG_UI_MERCHANT_PANEL
          :
        </strong>
        <br />
        <code>
           This is a message for the Merchant's Panel
        </code>
      </li>
      <li>
        <strong>
          MSG_SETTLEMENT_PAYER
          :
        </strong>
        <br />
        <code>
           This is a message for the customer
        </code>
      </li>
      <li>
        <strong>
          MSG_UI_PAYMENT_METHOD
          :
        </strong>
        <br />
        <code>
           This is a message for the payment method
        </code>
      </li>
      <li>
        <strong>
          PARAMS_OUT
          :
        </strong>
        <br />
        <code>
           ORDER_NUMBER,PAYMENT_ID,AMOUNT,CURRENCY,PAYMENT_METHOD,TIMESTAMP,STATUS
        </code>
      </li>
    </ul>
    <form
      action="https://payment.paytrail.com/e2"
      class="e2-test-form"
      method="post"
    >
      <input
        name="MERCHANT_ID"
        type="hidden"
        value="1"
      />
      <input
        name="ORDER_NUMBER"
        type="hidden"
        value="1"
      />
      <input
        name="PARAMS_IN"
        type="hidden"
        value="MERCHANT_ID,ORDER_NUMBER,PARAMS_IN,LOCALE,CURRENCY,REFERENCE_NUMBER,EXPIRATION_FOR_PAYMENT_CREATION,PAYMENT_METHODS,VAT_IS_INCLUDED,ALG,URL_SUCCESS,URL_CANCEL,URL_NOTIFY,ITEM_TITLE[0],ITEM_ID[0],ITEM_QUANTITY[0],ITEM_UNIT_PRICE[0],ITEM_VAT_PERCENT[0],ITEM_DISCOUNT_PERCENT[0],ITEM_TYPE[0],PAYER_PERSON_PHONE,PAYER_PERSON_EMAIL,PAYER_PERSON_FIRSTNAME,PAYER_PERSON_LASTNAME,PAYER_COMPANY_NAME,PAYER_PERSON_ADDR_STREET,PAYER_PERSON_ADDR_POSTAL_CODE,PAYER_PERSON_ADDR_TOWN,PAYER_PERSON_ADDR_COUNTRY,MSG_UI_MERCHANT_PANEL,MSG_SETTLEMENT_PAYER,MSG_UI_PAYMENT_METHOD,PARAMS_OUT"
      />
      <input
        name="LOCALE"
        type="hidden"
        value="fi_FI"
      />
      <input
        name="CURRENCY"
        type="hidden"
        value="EUR"
      />
      <input
        name="REFERENCE_NUMBER"
        type="hidden"
        value="RF123456789"
      />
      <input
        name="EXPIRATION_FOR_PAYMENT_CREATION"
        type="hidden"
        value="2021-01-01"
      />
      <input
        name="PAYMENT_METHODS"
        type="hidden"
        value="1,2"
      />
      <input
        name="VAT_IS_INCLUDED"
        type="hidden"
        value="0"
      />
      <input
        name="ALG"
        type="hidden"
        value="1"
      />
      <input
        name="URL_SUCCESS"
        type="hidden"
        value="https://example.com/success"
      />
      <input
        name="URL_CANCEL"
        type="hidden"
        value="https://example.com/cancel"
      />
      <input
        name="URL_NOTIFY"
        type="hidden"
        value="https://example.com/notify"
      />
      <input
        name="ITEM_TITLE[0]"
        type="hidden"
        value="Test Product"
      />
      <input
        name="ITEM_ID[0]"
        type="hidden"
        value="123"
      />
      <input
        name="ITEM_QUANTITY[0]"
        type="hidden"
        value="1"
      />
      <input
        name="ITEM_UNIT_PRICE[0]"
        type="hidden"
        value="100.00"
      />
      <input
        name="ITEM_VAT_PERCENT[0]"
        type="hidden"
        value="24"
      />
      <input
        name="ITEM_DISCOUNT_PERCENT[0]"
        type="hidden"
        value="10"
      />
      <input
        name="ITEM_TYPE[0]"
        type="hidden"
        value="1"
      />
      <input
        name="PAYER_PERSON_PHONE"
        type="hidden"
        value="040123456"
      />
      <input
        name="PAYER_PERSON_EMAIL"
        type="hidden"
        value="alice@example.org"
      />
      <input
        name="PAYER_PERSON_FIRSTNAME"
        type="hidden"
        value="Alice"
      />
      <input
        name="PAYER_PERSON_LASTNAME"
        type="hidden"
        value="Angel"
      />
      <input
        name="PAYER_COMPANY_NAME"
        type="hidden"
        value="Acme Ltd"
      />
      <input
        name="PAYER_PERSON_ADDR_STREET"
        type="hidden"
        value="Test Street 1"
      />
      <input
        name="PAYER_PERSON_ADDR_POSTAL_CODE"
        type="hidden"
        value="00100"
      />
      <input
        name="PAYER_PERSON_ADDR_TOWN"
        type="hidden"
        value="Helsinki"
      />
      <input
        name="PAYER_PERSON_ADDR_COUNTRY"
        type="hidden"
        value="FI"
      />
      <input
        name="MSG_UI_MERCHANT_PANEL"
        type="hidden"
        value="This is a message for the Merchant's Panel"
      />
      <input
        name="MSG_SETTLEMENT_PAYER"
        type="hidden"
        value="This is a message for the customer"
      />
      <input
        name="MSG_UI_PAYMENT_METHOD"
        type="hidden"
        value="This is a message for the payment method"
      />
      <input
        name="PARAMS_OUT"
        type="hidden"
        value="ORDER_NUMBER,PAYMENT_ID,AMOUNT,CURRENCY,PAYMENT_METHOD,TIMESTAMP,STATUS"
      />
      <input
        name="AUTHCODE"
        type="hidden"
        value="CA8A6BCB3A31EEB4114A6DE1A5DE8D48F7D33D3E0C081310E63CAB5A8E9E4CA1"
      />
      <button
        type="submit"
      >
        Pay
      </button>
    </form>
  </div>
</div>
`;

exports[`Form renders with minimum required fields 1`] = `
<div>
  <div>
    <form
      action="https://payment.paytrail.com/e2"
      class="e2-form"
      method="post"
    >
      <input
        name="MERCHANT_ID"
        type="hidden"
        value="1"
      />
      <input
        name="ORDER_NUMBER"
        type="hidden"
        value="1"
      />
      <input
        name="PARAMS_IN"
        type="hidden"
        value="MERCHANT_ID,ORDER_NUMBER,PARAMS_IN,LOCALE,CURRENCY,VAT_IS_INCLUDED,ALG,URL_SUCCESS,URL_CANCEL,URL_NOTIFY,ITEM_TITLE[0],ITEM_ID[0],ITEM_QUANTITY[0],ITEM_UNIT_PRICE[0],ITEM_VAT_PERCENT[0],ITEM_DISCOUNT_PERCENT[0],ITEM_TYPE[0],PAYER_PERSON_FIRSTNAME,PAYER_PERSON_LASTNAME,PARAMS_OUT"
      />
      <input
        name="LOCALE"
        type="hidden"
        value="fi_FI"
      />
      <input
        name="CURRENCY"
        type="hidden"
        value="EUR"
      />
      <input
        name="VAT_IS_INCLUDED"
        type="hidden"
        value="1"
      />
      <input
        name="ALG"
        type="hidden"
        value="1"
      />
      <input
        name="URL_SUCCESS"
        type="hidden"
        value="https://example.com/success"
      />
      <input
        name="URL_CANCEL"
        type="hidden"
        value="https://example.com/cancel"
      />
      <input
        name="URL_NOTIFY"
        type="hidden"
        value="https://example.com/notify"
      />
      <input
        name="ITEM_TITLE[0]"
        type="hidden"
        value="Test Product"
      />
      <input
        name="ITEM_ID[0]"
        type="hidden"
        value="123"
      />
      <input
        name="ITEM_QUANTITY[0]"
        type="hidden"
        value="1"
      />
      <input
        name="ITEM_UNIT_PRICE[0]"
        type="hidden"
        value="100.00"
      />
      <input
        name="ITEM_VAT_PERCENT[0]"
        type="hidden"
        value="24"
      />
      <input
        name="ITEM_DISCOUNT_PERCENT[0]"
        type="hidden"
        value="0"
      />
      <input
        name="ITEM_TYPE[0]"
        type="hidden"
        value="1"
      />
      <input
        name="PAYER_PERSON_FIRSTNAME"
        type="hidden"
        value="Alice"
      />
      <input
        name="PAYER_PERSON_LASTNAME"
        type="hidden"
        value="Angel"
      />
      <input
        name="PARAMS_OUT"
        type="hidden"
        value="ORDER_NUMBER,PAYMENT_ID,AMOUNT,CURRENCY,PAYMENT_METHOD,TIMESTAMP,STATUS"
      />
      <input
        name="AUTHCODE"
        type="hidden"
        value="94318EF18EAAA5B32A3F5484D8D8893ACFD4C0551BCDFBFDC52F1B7239F0CB1B"
      />
      <button
        type="submit"
      >
        Pay
      </button>
    </form>
  </div>
</div>
`;
